touchforms/backend/test_form_player_api.py

```diff
@@ -0,0 +1,233 @@
+
+
+CUR_DIR = os.path.dirname(__file__)
+
+
+class GUIStub(object):
+    def __getattr__(self, name):
+        return lambda _self: None
+
+
+class FormPlayerApiTest(unittest.TestCase):
+
+    def setUp(self):
+
+        try:
+            f = open(os.path.join(CUR_DIR, 'test_files/xforms/question_types.xml'), 'r')
+            self.xform = f.read()
+            f = open(os.path.join(CUR_DIR, 'test_files/xforms/groups.xml'), 'r')
+            self.group_xform = f.read()
+            f = open(os.path.join(CUR_DIR, 'test_files/xforms/repeat.xml'), 'r')
+            self.repeat_xform = f.read()
+            f = open(os.path.join(CUR_DIR, 'test_files/xforms/nested.xml'), 'r')
+            self.nested_xform = f.read()
+        except IOError:
+            self.fail('Could not read test form')
+
+        self.case = {
+            'case_id': 'legolas',
+            'properties': {
+                'case_type': 'dragon',
+                'case_name': 'rocky',
+                'date_opened': None,
+                'owner_id': 'ben-123',
+            },
+            'closed': False,
+            'indices': {},
+            'attachments': {},
+        }
+
+        self.form = os.path.join(CUR_DIR, 'test_files/simple_submission.xml')
+        self.restore = os.path.join(CUR_DIR, 'test_files/restores/simple_restore.xml')
+
+        self.session_data = {
+            'session_name': 'Village Healthe > Simple Form',
+            'app_version': '2.0',
+            'device_id': 'cloudcare',
+            'user_id': '51cd680c0bd1c21bb5e63dab99748248',
+            'additional_filters': {'footprint': True},
+            'domain': 'willslearningproject',
+            'host': 'http://localhost:8000',
+            'user_data': {},
+            'case_id_new_RegCase_0': '1c2e7c76f0c84eaea5b44bc7d1d3caf0',
+            'app_id': '6a48b8838d06febeeabb28c8c9516ab6',
+            'username': 'submission-test'
+        }
+
+        self.session_metadata = {
+            'extensions': [],
+            'session_data': self.session_data,
+            'nav_mode': 'fao',
+            'api_auth': {},
+            'staleness_window': 1,
+            'form_context': {
+                'all_case_ids': ['legolas', 'sauron'],
+                'case_model': self.case
+            },
+        }
+
+        self.commcare_instance = touchcare.perform_restore(
+            {},
+            self.session_data,
+            restore_xml=self.restore,
+        )
+
+
+        xformplayer._init()
+        self.manager = xformplayer.GlobalStateManager.get_globalstate()
+
+    def test_repeat_walking(self):
+
+        xform_session = xformplayer.XFormSession(
+            self.repeat_xform,
+            instance=None,
+            use_api=True,
+            **self.session_metadata
+        )
+
+        model = xform_session.fem
+        controller = xform_session.fec
+
+        assert(json_equals(xform_session.walk(), json.loads(WalkJson.walkToJson(model, controller))))
```

Comment:
i believe self.assertEquals should ensure that dicts are equal (maybe in 2.5 we need to use self.assertDictEquals)